code/datums/components/riding.dm

```diff
@@ -0,0 +1,441 @@
+
+/datum/component/riding
+	var/next_vehicle_move = 0 //used for move delays
+	var/vehicle_move_delay = 2 //tick delay between movements, lower = faster, higher = slower
+	var/keytype
+	var/atom/movable/_parent
+
+	var/slowed = FALSE
+	var/slowvalue = 1
+
+/datum/component/riding/Initialize()
+	. = ..()
+	if(!istype(parent, /atom/movable))
+		CRASH("RIDING DATUM ASSIGNED TO NON ATOM MOVABLE!")
+	_parent = parent
+	RegisterSignal(COMSIG_MOVABLE_BUCKLE, .proc/vehicle_mob_buckle)
+	RegisterSignal(COMSIG_MOVABLE_UNBUCKLE, .proc/vehicle_mob_unbuckle)
+	RegisterSignal(COMSIG_MOVABLE_MOVED, .proc/vehicle_moved)
+
+/datum/component/riding/proc/vehicle_mob_unbuckle(mob/living/M, force = FALSE)
+	restore_position(M)
+
+/datum/component/riding/proc/vehicle_mob_buckle(mob/living/M, force = FALSE)
+	handle_vehicle_offsets()
+
+/datum/component/riding/proc/handle_vehicle_layer()
+	if(_parent.dir != NORTH)
+		_parent.layer = ABOVE_MOB_LAYER
+	else
+		_parent.layer = OBJ_LAYER
+
+/datum/component/riding/proc/vehicle_moved()
+	for(var/mob/living/M in _parent.buckled_mobs)
+		ride_check(M)
+	handle_vehicle_offsets()
+	handle_vehicle_layer()
+
+/datum/component/riding/proc/ride_check(mob/living/M)
+	return TRUE
+
+/datum/component/riding/proc/force_dismount(mob/living/M)
+	_parent.unbuckle_mob(M)
+
+/datum/component/riding/proc/handle_vehicle_offsets()
+	var/_parent_dir = "[_parent.dir]"
+	var/passindex = 0
+	if(_parent.has_buckled_mobs())
+		for(var/m in _parent.buckled_mobs)
+			passindex++
+			var/mob/living/buckled_mob = m
+			var/list/offsets = get_offsets(passindex)
+			var/rider_dir = get_rider_dir(passindex)
+			buckled_mob.setDir(rider_dir)
+			dir_loop:
+				for(var/offsetdir in offsets)
+					if(offsetdir == _parent_dir)
+						var/list/diroffsets = offsets[offsetdir]
+						buckled_mob.pixel_x = diroffsets[1]
+						if(diroffsets.len >= 2)
+							buckled_mob.pixel_y = diroffsets[2]
+						if(diroffsets.len == 3)
+							buckled_mob.layer = diroffsets[3]
+						break dir_loop
+
+//Override this to set your vehicle's various pixel offsets
+/datum/component/riding/proc/get_offsets(pass_index) // list(dir = x, y, layer)
+	return list("[NORTH]" = list(0, 0), "[SOUTH]" = list(0, 0), "[EAST]" = list(0, 0), "[WEST]" = list(0, 0))

```

Comment:
There are defines for these in space_transition.dm, move them up to __DEFINES/maps.dm and use them

Reply:
fixed